docs/lara-plugin-api.md

```diff
@@ -0,0 +1,129 @@
+
+# LARA Plugin API
+
+A LARA Plugin architecture which can be used by
+javascript developers to add new features to LARA.
+
+NP 2018-08-29 -- This documentation is being developed and is
+subject to change.  
+
+## The hosted LARA Plugin API
+
+This describes the functions attached to the global `window.LARA` object
+LARA exports.
+
+```typescript
+
+/****************************************************************************
+// @function addSidebar: Ask lara to add a new sidebar window
+// @arg {ISidebarOptions} sidebarOptions
+// @returns A parent DomNode where component can render content
+****************************************************************************/
+LARA.addSidebar(sidebarOptions: ISidebarOptions) : DomNode
+
+
+/****************************************************************************
+@function openPopup: Ask lara to add a new popup window
+@arg {IPopupOptions} popupOptions
+@returns A parent DomNode node where component can render content
+****************************************************************************/
+LARA.openPopup(popupOptions: IPopupOptions) : DomNode
```

Comment:
I think @scytacki was voting for keeping everything async or Promise-based.
Personally, I'd keep it like that - simple. If for some reason this needs to be async in the future, we can create a new method and/or update API. Making **everything** returning promises would make the plugin code pretty strange and painful to work with. We'll have tons of nested `.then()` statements or async/await thingies. 
That's my gut feeling / opinion, but I remember this note from @scytacki, so we can use this thread to discuss it.

Reply:
On a related note, I'd rather return a hash (e.g. `{ container: HTMLElement }`) even if we can return basic type. It means that we can extend this API and response without being backward incompatible. And in this case, it's pretty likely that we might provide some handles to close the popup or reposition it, and so on.

Reply:
Would this alternative work:
The IPopupOptions contains a React Class name, and a set of properties.  Then openPopup doesn't need to return anything.  And this could be used by something in an iframe or worker using a LARA provided iframe component, with a property which was a url to the iframe content.

Reply:
This means plugins need to provide react components, but I'm OK with that.

Reply:
Hm, passing React component class (or class name) and set of properties feels a bit opinionated to me. I'd rather pass any HTML element. The plugin can prepare this HTML element whatever way it wants. Using document.createElement, using jQuery, or using React.

Actually, that was my very first proposal. However, after discussing that with @knowuh, we thought that React would fail when we try to render a component into a "detached" HTML container (not part of the page tree yet, as it will be done by LARA later). But I've just tried that and it all works fine:

```typescript
// That's the plugin part:
const container = document.createElement("div");
ReactDOM.render(
  <GlossaryPopup
    word="eardrum"
    definition="An eardrum is a membrane, or thin piece of skin, stretched tight like a drum."
    askForUserDefinition={true}
    onUserDefinitionsUpdate={newUserDefinition}
  />,
  container
);
// now, send container to LARA, e.g.:
// LARA.openPopup({ content: container });
// (...)
// Later, done by LARA:
document.body.appendChild(container);
// it works ;)
```

There's one small drawback of the rendering HTML when elements are detached. You can't measure elements, as they all have 0 width and height (e.g. if you need to measure some dynamic text label). Lab works exactly that way. Lab components use jQuery to render themselves and later some main controller adds them into right spot. This measurment problem is the only one I had while doing it. However, it's very rare that you need it. And even if you do, you can always attach your element to body, measure, and detach (that's what I did in Lab when I had).

So, if we want to make this API simpler, I'd vote for this approach.

Reply:
I liked Piotr's sample code, so I have extended his example, trying to pseudo-code a complete implementation, using two different `openPopup` signatures, one in which we provide a DomNode, and one where we get a DomNode back.

These basically look identical to me. The question is are we more worried about handling async return values from `openPopup()` or are we more worried about edge cases in layout behavior when we need to measure things?  I guess I think both are low-risk, and don't care much either way.

1. Pass in the content to the call to openPopup:

```
const definitions = {
  'word': {text: 'word definition'},
  'eardrum': {text: 'An eardrum is a membrane, or thin piece of skin, stretched tight like a drum.'}
  // …
}

const openPopup = (event) => {
  // hand-waving …
  const word = event.target.innerText
  const definition = definitions[word].text
  const {x,y} = event.target.position;

  const container = document.createElement("div")

  ReactDOM.render(
    <GlossaryPopup
      word={word}
      definition={definition}
      askForUserDefinition={true}
      onUserDefinitionsUpdate={newUserDefinition}
    />,
    container
  );

  const options = {
    title: "Glossary",
    position: {screenX: x, screenY: y},
    content: container,
    draggable: true,
    onClose: () =>  { /* ? …  */ }
  }

  LARA.openPopup(options)
}

const words = Object.keys(definitions)
const replace = '<span data-decorationtype="glossary-definition>$1</span>"'
LARA.decorateContent({words: words, replace: replace, listeners: [{wordClick: openGlossaryPopup}]})

```

2. Get the container from the return value of openPopup:

```

const definitions = {
  'word': {text: 'word definition'},
  'eardrum': {text: 'An eardrum is a membrane, or thin piece of skin, stretched tight like a drum.'}
  // …
}

const openGlossaryPopup = (event) => {

  // hand-waving …
  const {x,y} = event.target.position
  const word = event.target.innerText
  const definition = definitions[word].text

  const options = {
    title: "Glossary",
    position: {screenX: x, screenY: y},
    draggable: true,
    onClose: () =>  { /* ? …  */ }
  }

  const container = LARA.openPopup(options)

  ReactDOM.render(
    <GlossaryPopup
      word={word}
      definition={definition}
      askForUserDefinition={true}
      onUserDefinitionsUpdate={newUserDefinition}
    />,
    container
  );


}

const words = Object.keys(definitions)
const replace = '<span data-decorationtype="glossary-definition>$1</span>"'
LARA.decorateContent({words: words, replace: replace, listeners: [{wordClick: openGlossaryPopup}]})

```

Reply:
I agree there's no significant difference between them, and I'm fine with both.

I'd say I'd slightly prefer option 1, as content goes together with other options like title, position, etc. So, everything you need to do to open a popup is contained in this one options hash.

The second approach is a bit spread out. You set title and position using options, but then render content using returned element. So, you need to remember what goes where. It'd be more visible if there're more containers returned - e.g. a container for the icon, header container, and so on.

Also, both options can exist together if we ever need it, so API can be easily extended either way.

Reply:
Let us go with option 1 then.  I was against it at first, but have been won over.

But one other thing I am generally worried about (which is probably what @scytacki  was trying to warn us about) is that we can have nested mounted React components.  So if the popup container ever becomes a child of a React mounted component, we might be in trouble. 

Is this something we should be worried about, or can we ignore it for now?

Reply:
I haven't read the discussion above carefully. But here is my 2 cents:

I was mainly trying to find a way so it would be easy to use the same API from an iframe or worker. So having a hard rule that everything needs to async, would help with that.  But it also means that you can't pass dom elements and listeners around.  

In addition to this it would be useful to plan for a fully React LARA UI that was a SPA.  This is where I'd like LARA to go. And it would be great if we could get there by pulling out components from LARA and making them interactives or plugins.  And then those same interactives and plugins could work both in the SPA LARA and the current LARA.  It sounds like from the last comment that, this would not work because the glossary could not embed itself inside of a already formed React tree.

But given all of this, I think it is best to get started building option 1 as you decided. And if we have time we can consider a refactor that sets us up for both iframes and a React SPA.

Reply:
I think option 1 and 2 can exist together. React + props could be handled too. I don't think it would complicate LARA and plugins will have a choice.

> I was mainly trying to find a way so it would be easy to use the same API from an iframe or worker.

This still seems impossible with React component name + props approach if I understand you correctly. We can't pass component class thorugh iframe, and even if we pass name as a string, this component implementation won't be available on the LARA side.

The only option I can see here is that iframe plugin could send content as a string like ```{ content: "<iframe src='some-component.concord.org/view-a.html'/>" }``` and LARA could interpret this string (trivial using jQuery). Actually, it would be yet another option that LARA could support without much work.

> So if the popup container ever becomes a child of a React mounted component, we might be in trouble.

That's a good question. It feels like a possible problem. However, I don't know if it would break things or not. I guess it should be pretty easy to test and see what happens.